Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+        fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) * fpu_get(0));
+    } else {
+        auto new_f64 = insn.modrm().read64(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f64 = bit_cast<double>(new_f64.value());
+        fpu_set(0, fpu_get(0) * f64);
+    }
+}
+void SoftFPU::FMULP(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) * fpu_get(0));
+    fpu_pop();
+}
+
+void SoftFPU::FIMUL_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m32int = (i32)insn.modrm().read32(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    fpu_set(0, fpu_get(0) * (long double)m32int);
+}
+void SoftFPU::FIMUL_RM16(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m16int = (i16)insn.modrm().read16(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    fpu_set(0, fpu_get(0) * (long double)m16int);
+}
+
+void SoftFPU::FDIV_RM32(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(0, fpu_get(0) / fpu_get(insn.modrm().register_index()));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f32 = bit_cast<float>(new_f32.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, fpu_get(0) / f32);
+    }
+}
+void SoftFPU::FDIV_RM64(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) / fpu_get(0));
+    } else {
+        auto new_f64 = insn.modrm().read64(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f64 = bit_cast<double>(new_f64.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, fpu_get(0) / f64);
+    }
+}
+void SoftFPU::FDIVP(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+    fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) / fpu_get(0));
+    fpu_pop();
+}
+
+void SoftFPU::FDIVR_RM32(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(0, fpu_get(insn.modrm().register_index()) / fpu_get(0));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f32 = bit_cast<float>(new_f32.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, f32 / fpu_get(0));
+    }
+}
+void SoftFPU::FDIVR_RM64(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        // XXX this is FDIVR, Instruction decodes this weirdly
```

Comment:
XXX -> FIXME? Very odd to see just this one being different.

Reply:
Not mine,
Looking at it closely its the wrong way around...

Reply:
but switching breaks Cube even more, so either its so broken it looks good again, or https://www.felixcloutier.com/x86/fdiv:fdivp:fidiv is wrong, and that took it from the manual